### DISCUSSION TITLE
How to inject data/services into the Core project from platform-specific projects?

### DISCUSSION THREAD

**Discussion by Ricent82** [AvaloniaUI/Avalonia#13890] [Q&A] [answered]
I'm trying to figure out how to inject something (i.e. data, a service, etc) into the Core project from the platform-specific projects. My guess is that it has something to do with the AppBuilder, and should come in i.e. as an arg to Initialize() - but I can't seem to find any documentation relating to this, nor have I had any luck searching the discussions. Any ideas?

Thanks in advance.

**Comment by Ricent82** [comment]
Nevermind, just figured it out right after I posted that.  For anyone else who's looking:

1. Create a public static field in App.axaml.cs (in the Core project)
2. Then in the platform specific project (i.e. the Android), in MainActivity.cs 's `CustomizeAppBuilder()`, you can just set App.TheField to whatever you want.

**Reply by maxkatz6** [reply]
It's one of the ways. A bit more traditional would be to use IoC/DI.
Or, alternatively, you can add multiple target frameworks to the Core project, and differentiate them by `#if ANDROID || IOS` , like MAUI does it.

**Reply by Ricent82** [reply]
Thanks for the reply. Could you provide a little more specifics on how the IoC/DI approach would be done? I couldn't find any examples.

Thanks again

**Reply by IanRawley** [reply, marked as answer]
1. Have the platform independent project define an interface for the service you need.
2. Have the platform specific projects define a class implementing that interface, then register it with your dependency injection method of choice.  Out of the box for Avalonia would be Splat, e.g. in MainActivity.cs (Android) have the following or similar according to your needs.
```Locator.CurrentMutable.RegisterConstant<IPlatformCapabilitiesService>(platformCapabilities);```
3.  Have the platform independent project request the service from Splat when needed, for example as follows:
```_platformCapabilitiesService = Locator.Current.GetService<IPlatformCapabilitiesService>();```

**Reply by Ricent82** [reply]
Ah, gotcha. 1 and the first clause of 2 is exactly what I did (the interface & platform-specific classes). I just set the App's static field to each class instance directly, rather than using Splat. Pretty close tho :)
